test/controllers/tasks_controller_test.rb

```diff
@@ -105,24 +115,77 @@
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
     it "can update an existing task" do
-      skip
-      # Your code here
+      update_task_hash = {
+        task: {
+          name: "update task",
+          description: "new task description",
+          completion_date: "today",
+        },
+      }
+      existing_task = Task.create(name: "Hiking", description: "fun", completion_date: "Friday")
+
+      patch task_path(existing_task), params: update_task_hash
+
+      existing_task.reload
+      expect(existing_task.completion_date).must_equal "today"
     end
 
     it "will redirect to the root page if given an invalid id" do
-      skip
-      # Your code here
+      # Arrange
+      update_task_hash = {
+        task: {
+          name: "update task",
+          description: "new task description",
+          completion_date: "today",
+        },
+      }
+      invalid_task_id = "Invalid"
+      #Act
+      patch task_path(invalid_task_id), params: update_task_hash
+      # Assert
+      must_redirect_to root_path
     end
   end
 
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    it "returns a 404 if the book is not found" do
+      # Arrange
+      invalid_task_id = "NOT A VALID ID"
+
+      # Try to do the Books# destroy action
+      expect {
+        # Act
+        delete task_path(invalid_task_id)
+      }.must_change "Task.count", 0
+      must_respond_with :not_found
+    end
 
+    it "can delete a book" do
+      # Arrange - Create a book
+      new_task = Task.create(name: "Study")
+
+      expect {
+        # Act
+        delete task_path(new_task.id)
+        # Assert
+      }.must_change "Task.count", -1
+
+      must_respond_with :redirect
+      must_redirect_to tasks_path
+    end
   end
 
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    it "can mark a task as complete" do
+      # Arrange
+      complete_task = Task.create(name: "Study", description: "At Ada", completion_date: "Sunday", status: false)
+      # Act
+      patch toggle_path(complete_task.id)
+      toggle = Task.find_by(id: complete_task.id)
+      # Assert
+      expect(toggle.status).must_equal true
```

Comment:
This test is failing... Also you need to test that once you toggle a task to complete, you can toggle it incomplete.